[te] Add a benchmark harness

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45906 [te][llvm] Enable fused multiply-add (fma) in code generation
* #45905 [te] Tiled (m=32 x n=32) gemm benchmark
* **#45875 [te] Add a benchmark harness**
* #45514 [te] Add a 2D convolution example test

Summary:

Adds a googlebenchmark harness for perf testing programs generated by
tensorexpr, sans any pytorch wrappings (for python-level benchmarks of
tensorexpr, see benchmarks/tensorexpr).

Currently there's a harness for gemm that sets up the problem using torch (and
also measures the perf of a torch::mm to give a baseline).

Right now there's just an unoptimized implementation that is expected to be not
very fast.  More optimized versions are coming.

Sample output from my dev box:
```
Run on (48 X 2501 MHz CPU s)
CPU Caches:
  L1 Data 32K (x24)
  L1 Instruction 32K (x24)
  L2 Unified 256K (x24)
  L3 Unified 30720K (x2)
--------------------------------------------------------------------------------------------
Benchmark                                     Time           CPU Iterations UserCounters...
--------------------------------------------------------------------------------------------
Gemm/Torch/128/128/128                    73405 ns      73403 ns       8614 GFLOPS=57.1411G/s
Gemm/TensorExprNoopt/128/128/128        3073003 ns    3072808 ns        229 GFLOPS=1.36497G/s
```

Differential Revision: [D24142403](https://our.internmc.facebook.com/intern/diff/D24142403)

bertmaher committed:
Add a TE benchmark harness

[ghstack-poisoned]

zheng-xq commented on benchmarks/cpp/tensorexpr/tensorexpr.cpp:
```diff
@@ -0,0 +1,67 @@
+  }
+
+  void TearDown(benchmark::State& state) {
+    state.counters["GFLOPS"] =
```
Could you post an example output in the commit message?

zheng-xq commented on benchmarks/cpp/tensorexpr/tensorexpr.cpp:
```diff
@@ -0,0 +1,67 @@
+
+BENCHMARK_DEFINE_F(Gemm, Torch)(benchmark::State& state) {
+  for (auto _ : state) {
+    torch::mm_out(C, A, B);
```
Minor: This is fine as it is. But this default of this benchmark completely fits in L2, which might not be representative. On the other hand, flushing cache might seem too big an interruption.

A possible middle-ground, is to have vectors of A, B and C, and 
```
mm_out(A[i % size_a], B[2 * i % size_b], C[3 * i % size_c]). 
```

bertmaher replied:
I definitely agree we want to have more problem sizes.  I started with 128x128x128 exactly because it fits in L2 :)

zheng-xq commented on benchmarks/cpp/tensorexpr/tensorexpr.cpp:
```diff
@@ -0,0 +1,67 @@
+  loop.prepareForCodegen();
+  te::Stmt* s = loop.root_stmt();
+  s = te::IRSimplifier::simplify(s);
+  auto cg = CreateCodeGen("llvm_codegen", s, {AP, BP, CT});
```
BTW, can we check for functional correctness on the first iteration outside the measured range?

ZolotukhinM commented:
Nit: might be good to have a more verbose description of what benchmarks we're harnessing here, and where and how to add new ones, etc.

bertmaher committed:
Update on "[te] Add a benchmark harness"


Sample output from my dev box:
```
Run on (48 X 2501 MHz CPU s)
CPU Caches:
  L1 Data 32K (x24)
  L1 Instruction 32K (x24)
  L2 Unified 256K (x24)
  L3 Unified 30720K (x2)
--------------------------------------------------------------------------------------------
Benchmark                                     Time           CPU Iterations UserCounters...
--------------------------------------------------------------------------------------------
Gemm/Torch/128/128/128                    73405 ns      73403 ns       8614 GFLOPS=57.1411G/s
Gemm/TensorExprNoopt/128/128/128        3073003 ns    3072808 ns        229 GFLOPS=1.36497G/s
```

[ghstack-poisoned]

bertmaher committed:
Update on "[te] Add a benchmark harness"


Summary:

Adds a googlebenchmark harness for perf testing programs generated by
tensorexpr, sans any pytorch wrappings (for python-level benchmarks of
tensorexpr, see benchmarks/tensorexpr).

Currently there's a harness for gemm that sets up the problem using torch (and
also measures the perf of a torch::mm to give a baseline).

Right now there's just an unoptimized implementation that is expected to be not
very fast.  More optimized versions are coming.

Sample output from my dev box:
```
Run on (48 X 2501 MHz CPU s)
CPU Caches:
  L1 Data 32K (x24)
  L1 Instruction 32K (x24)
  L2 Unified 256K (x24)
  L3 Unified 30720K (x2)
--------------------------------------------------------------------------------------------
Benchmark                                     Time           CPU Iterations UserCounters...
--------------------------------------------------------------------------------------------
Gemm/Torch/128/128/128                    73405 ns      73403 ns       8614 GFLOPS=57.1411G/s
Gemm/TensorExprNoopt/128/128/128        3073003 ns    3072808 ns        229 GFLOPS=1.36497G/s
```

[ghstack-poisoned]

bwasti commented on caffe2/CMakeLists.txt:
```diff
@@ -1261,6 +1261,12 @@ if(BUILD_STATIC_RUNTIME_BENCHMARK)
   target_link_libraries(static_runtime_test torch_library gtest_main)
 endif()
 
+if(BUILD_TENSOREXPR_BENCHMARK)
```
just a heads up that if I enable this and BUILD_STATIC_RUNTIME_BENCHMARK I get into a weird state where cmake complains

```
CMake Error at caffe2/CMakeLists.txt:1265 (add_subdirectory):
  The binary directory

    /mnt/ssd1/bwasti/pytorch/build/bin

  is already used to build a source directory.  It cannot be used to build                                                                                                                                                                      source directory                                                                                                                                                                                                                                                                                                                                                                                                                                                                              /mnt/ssd1/bwasti/pytorch/benchmarks/cpp/tensorexpr
```

bertmaher replied:
Thanks, fixed it!

bertmaher committed:
Update on "[te] Add a benchmark harness"


Summary:

Adds a googlebenchmark harness for perf testing programs generated by
tensorexpr, sans any pytorch wrappings (for python-level benchmarks of
tensorexpr, see benchmarks/tensorexpr).

Currently there's a harness for gemm that sets up the problem using torch (and
also measures the perf of a torch::mm to give a baseline).

Right now there's just an unoptimized implementation that is expected to be not
very fast.  More optimized versions are coming.

Sample output from my dev box:
```
Run on (48 X 2501 MHz CPU s)
CPU Caches:
  L1 Data 32K (x24)
  L1 Instruction 32K (x24)
  L2 Unified 256K (x24)
  L3 Unified 30720K (x2)
--------------------------------------------------------------------------------------------
Benchmark                                     Time           CPU Iterations UserCounters...
--------------------------------------------------------------------------------------------
Gemm/Torch/128/128/128                    73405 ns      73403 ns       8614 GFLOPS=57.1411G/s
Gemm/TensorExprNoopt/128/128/128        3073003 ns    3072808 ns        229 GFLOPS=1.36497G/s
```

Differential Revision: [D24142403](https://our.internmc.facebook.com/intern/diff/D24142403)

[ghstack-poisoned]

bertmaher committed:
Update on "[te] Add a benchmark harness"


Summary:

Adds a googlebenchmark harness for perf testing programs generated by
tensorexpr, sans any pytorch wrappings (for python-level benchmarks of
tensorexpr, see benchmarks/tensorexpr).

Currently there's a harness for gemm that sets up the problem using torch (and
also measures the perf of a torch::mm to give a baseline).

Right now there's just an unoptimized implementation that is expected to be not
very fast.  More optimized versions are coming.

Sample output from my dev box:
```
Run on (48 X 2501 MHz CPU s)
CPU Caches:
  L1 Data 32K (x24)
  L1 Instruction 32K (x24)
  L2 Unified 256K (x24)
  L3 Unified 30720K (x2)
--------------------------------------------------------------------------------------------
Benchmark                                     Time           CPU Iterations UserCounters...
--------------------------------------------------------------------------------------------
Gemm/Torch/128/128/128                    73405 ns      73403 ns       8614 GFLOPS=57.1411G/s
Gemm/TensorExprNoopt/128/128/128        3073003 ns    3072808 ns        229 GFLOPS=1.36497G/s
```

Differential Revision: [D24142403](https://our.internmc.facebook.com/intern/diff/D24142403)

[ghstack-poisoned]
